plugins/janus_nosip.c

```diff
@@ -0,0 +1,2213 @@
+			GList *sl = old_sessions;
+			JANUS_LOG(LOG_HUGE, "Checking %d old NoSIP sessions...\n", g_list_length(old_sessions));
+			while(sl) {
+				janus_nosip_session *session = (janus_nosip_session *)sl->data;
+				if(!session) {
+					sl = sl->next;
+					continue;
+				}
+				if (now-session->destroyed >= 5*G_USEC_PER_SEC) {
+					/* We're lazy and actually get rid of the stuff only after a few seconds */
+					JANUS_LOG(LOG_VERB, "Freeing old NoSIP session\n");
+					GList *rm = sl->next;
+					old_sessions = g_list_delete_link(old_sessions, sl);
+					sl = rm;
+					g_free(session->info);
+					session->info = NULL;
+					janus_sdp_free(session->sdp);
+					session->sdp = NULL;
+					g_free(session->media.remote_ip);
+					session->media.remote_ip = NULL;
+					janus_nosip_srtp_cleanup(session);
+					session->handle = NULL;
+					g_free(session);
+					session = NULL;
+					continue;
+				}
+				sl = sl->next;
+			}
+		}
+		janus_mutex_unlock(&sessions_mutex);
+		g_usleep(500000);
+	}
+	JANUS_LOG(LOG_INFO, "NoSIP watchdog stopped\n");
+	return NULL;
+}
+
+
+static void janus_nosip_detect_local_ip(char *buf, size_t buflen) {
+	JANUS_LOG(LOG_VERB, "Autodetecting local IP...\n");
+
+	struct sockaddr_in addr;
+	socklen_t len;
+	int fd = socket(AF_INET, SOCK_DGRAM, 0);
+	if (fd == -1)
+		goto error;
+	addr.sin_family = AF_INET;
+	addr.sin_port = htons(1);
+	inet_pton(AF_INET, "1.2.3.4", &addr.sin_addr.s_addr);
+	if (connect(fd, (const struct sockaddr*) &addr, sizeof(addr)) < 0)
+		goto error;
+	len = sizeof(addr);
+	if (getsockname(fd, (struct sockaddr*) &addr, &len) < 0)
+		goto error;
+	if (getnameinfo((const struct sockaddr*) &addr, sizeof(addr),
+			buf, buflen,
+			NULL, 0, NI_NUMERICHOST) != 0)
+		goto error;
+	close(fd);
+	return;
+
+error:
+	if (fd != -1)
+		close(fd);
+	JANUS_LOG(LOG_VERB, "Couldn't find any address! using 127.0.0.1 as the local IP... (which is NOT going to work out of your machine)\n");
+	g_strlcpy(buf, "127.0.0.1", buflen);
+}
+
+
+/* Plugin implementation */
+int janus_nosip_init(janus_callbacks *callback, const char *config_path) {
+	if(g_atomic_int_get(&stopping)) {
+		/* Still stopping from before */
+		return -1;
+	}
+	if(callback == NULL || config_path == NULL) {
+		/* Invalid arguments */
+		return -1;
+	}
+
+	/* Read configuration */
+	char filename[255];
+	g_snprintf(filename, 255, "%s/%s.cfg", config_path, JANUS_NOSIP_PACKAGE);
+	JANUS_LOG(LOG_VERB, "Configuration file: %s\n", filename);
+	janus_config *config = janus_config_parse(filename);
+	if(config != NULL) {
+		janus_config_print(config);
+
+		gboolean local_ip_set = FALSE;
+		janus_config_item *item = janus_config_get_item_drilldown(config, "general", "local_ip");
+		if(item && item->value) {
+			int family;
+			if (!janus_is_ip_valid(item->value, &family)) {
+				JANUS_LOG(LOG_WARN, "Invalid local IP specified: %s, guessing the default...\n", item->value);
+			} else {
+				/* Verify that we can actually bind to that address */
```

Comment:
As I told Saul, this has already happened... #776 by @cmacq2 adds this and more. It's just that I'm lazy and haven't updated the code we have around that does similar things to use that instead :wink: